substrate/frame/identity/src/lib.rs

```diff
@@ -1164,116 +1219,184 @@ pub mod pallet {
+			let username_info =
+				UsernameInfoOf::<T>::take(&username).ok_or(Error::<T>::NoUsername)?;
+			// If this is the primary username, remove the entry from the account -> username map.
+			UsernameOf::<T>::mutate(&username_info.owner, |maybe_primary| {
+				if match maybe_primary {
+					Some(primary) if *primary == username => true,
+					_ => false,
+				} {
+					*maybe_primary = None;
+				}
+			});
+			match username_info.provider {
+				Provider::AuthorityDeposit(username_deposit) => {
+					let suffix =
+						Self::suffix_of_username(&username).ok_or(Error::<T>::InvalidUsername)?;
+					if let Some(authority_account) =
+						AuthorityOf::<T>::get(&suffix).map(|auth_info| auth_info.account_id)
+					{
+						let err_amount =
+							T::Currency::unreserve(&authority_account, username_deposit);
+						debug_assert!(err_amount.is_zero());
+					}
+				},
+				Provider::Allocation => {
+					// We don't refund the allocation, it is lost.
+				},
+				Provider::System => return Err(Error::<T>::InvalidTarget.into()),
+			}
+			Self::deposit_event(Event::UsernameRemoved { username });
+			Ok(Pays::No.into())
+		}
 
-	/// Get the subs of an account.
-	pub fn subs(who: &T::AccountId) -> Vec<(T::AccountId, Data)> {
-		SubsOf::<T>::get(who)
-			.1
-			.into_iter()
-			.filter_map(|a| SuperOf::<T>::get(&a).map(|x| (a, x.1)))
-			.collect()
+		/// Call with [ForceOrigin](crate::Config::ForceOrigin) privileges which deletes a username
+		/// and slashes any deposit associated with it.
+		#[pallet::call_index(24)]
+		#[pallet::weight(T::WeightInfo::kill_username(0))]
+		pub fn kill_username(
+			origin: OriginFor<T>,
+			username: Username<T>,
+		) -> DispatchResultWithPostInfo {
+			T::ForceOrigin::ensure_origin(origin)?;
+			let username_info =
+				UsernameInfoOf::<T>::take(&username).ok_or(Error::<T>::NoUsername)?;
+			// If this is the primary username, remove the entry from the account -> username map.
+			UsernameOf::<T>::mutate(&username_info.owner, |maybe_primary| {
+				if match maybe_primary {
+					Some(primary) if *primary == username => true,
+					_ => false,
+				} {
+					*maybe_primary = None;
+				}
+			});
+			let _ = UnbindingUsernames::<T>::take(&username);
+			let actual_weight = match username_info.provider {
+				Provider::AuthorityDeposit(username_deposit) => {
+					let suffix =
+						Self::suffix_of_username(&username).ok_or(Error::<T>::InvalidUsername)?;
+					if let Some(authority_account) =
+						AuthorityOf::<T>::get(&suffix).map(|auth_info| auth_info.account_id)
+					{
+						T::Slashed::on_unbalanced(
+							T::Currency::slash_reserved(&authority_account, username_deposit).0,
+						);
+					}
+					T::WeightInfo::kill_username(0)
+				},
+				Provider::Allocation => {
+					// We don't refund the allocation, it is lost, but we do refund some weight.
+					T::WeightInfo::kill_username(1)
+				},
+				Provider::System => {
+					// Force origin can remove system usernames.
+					T::WeightInfo::kill_username(1)
```

Comment:
This could be used in big batches by governance to remove loads of usernames granted by a bad authority. My reasoning was that when we scale like that, small weight discrepancies are also scaled, so I wanted to be more accurate with this.